File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/user/SaveUserUseCase.java
package com.lattis.ellipse.domain.interactor.user;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.model.User;
import com.lattis.ellipse.domain.repository.UserRepository;

import io.reactivex.Observable;

public class SaveUserUseCase extends UseCase<User> {

    private UserRepository userRepository;
    private User user;

    protected SaveUserUseCase(ThreadExecutor threadExecutor,
                              PostExecutionThread postExecutionThread,
                              UserRepository userRepository) {
        super(threadExecutor, postExecutionThread);
        this.userRepository = userRepository;
    }

    public SaveUserUseCase withValue(User user) {
        this.user = user;
        return this;
    }

    @Override
    protected Observable<User> buildUseCaseObservable() {
        return userRepository.saveUser(user);
    }
}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/user/SaveUserUseCase.kt
package com.lattis.domain.usecase.user

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.UserRepository
import com.lattis.domain.usecase.base.UseCase
import com.lattis.domain.models.User
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class SaveUserUseCase @Inject constructor(
    threadExecutor: ThreadExecutor,
    postExecutionThread: PostExecutionThread,
    private val userRepository: UserRepository
) : UseCase<User>(threadExecutor, postExecutionThread) {
    private var user: User? = null
    fun withUser(user: User): SaveUserUseCase {
        this.user = user
        return this
    }

    override fun buildUseCaseObservable(): Observable<User> {
        return userRepository.saveUser(user!!)
    }

}